Add blog post about v03x/v04x migration

lgierth opened the pull request:
I think this has all the essential info. @RichardLitt @jbenet thoughts?

lgierth commented:
I'm not even sure a blog post is the perfect medium, but we can certainly use it to discuss and ship

RichardLitt commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,56 @@
+## The public gateways at ipfs.io
+
+For the time being, https://ipfs.io uses both v03x and v04x to service requests.
+Which ever responds successfully (2xx or 3xx status code) first,
```
whichever

RichardLitt commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,56 @@
+
+For the time being, https://ipfs.io uses both v03x and v04x to service requests.
+Which ever responds successfully (2xx or 3xx status code) first,
+gets to serve. The other response is discarded.
```
gets to serve its content

RichardLitt commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,56 @@
+These domain names will stay around as long as the respective network is
+supported by the public gateway.
+
+Expect the v03x gateway to be **supported until end of April 2016.**
```
until the end

RichardLitt commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,56 @@
+
+## The default bootstrappers
+
+These are the 8 bootstrap nodes in go-ipfs's default config.
```
Which are? Should they be posted here? How do I see the default nodes?

RichardLitt commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,56 @@
+but are available at default bootstrappers only to one.
+
+We'll gradually shift the v03x/v04x ratio to v04x.
+Expect the v03x bootstrappers to be **supported until end of April 2016.**
```
until the end

RichardLitt commented:
Overall, LGTM. However: what does this mean for me? Do I need to update my IPFS daemon? Do I need to update my server that uses IPFS? Something to clarify that would be good, I think.

lgierth commented:
Another thing that comes to mind is the question of whether v03x receives further patches/updates. It doesn't right now and it's not planned, but we should probably say so. cc @jbenet @whyrusleeping

lgierth commented:
Addressed the feedback and open questions, and generally improved the thing. It's pretty good now I think.

- Preview: https://ipfs.io/ipfs/QmQUuZSSYJRn8sXBDsUfyoWbkD9BS4W7UNXFsAdtRVTwSy/9-v04x-migration/
- Diff: https://github.com/ipfs/blog/commit/26c22b850e67d9955cdeb317d8fa0406ea84748c

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+---
+
+Good news everyone! We are about to release go-ipfs 0.4.0,
+which contains lots of great changes and enhancements.
```
i think we want a separate post about "the great changes and enhancements". can anyone pick this up? maybe @RichardLitt with help of @whyrusleeping and @diasdavid?

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+This means that there will be, and already are, two separate networks.
+We'll call these networks v03x and v04x.
+IPFS nodes built from the master branch are already part of v04x.
+The Docker image `ipfs/go-ipfs` is built from master and thus also part of v04x.
```
we should have the docker image follow the tags i have in `jbenet/go-ipfs`:
- `hub.docker.com/u/ipfs/go-ipfs@release` follows `github.com/ipfs/go-ipfs@release` (0.3.11 now)
- `hub.docker.com/u/ipfs/go-ipfs@latest` follows `github.com/ipfs/go-ipfs@release`  (0.3.11 now)
- `hub.docker.com/u/ipfs/go-ipfs@master` follows `github.com/ipfs/go-ipfs@master` (0.4.0 now)

lgierth replied:
We'll be able to have these only after the next release. It actually had a v0.3.11 tag until I removed it today (ipfs/go-ipfs#1944). The problem is that the dockerfile on 0.3.x just downloads `ipfs_master_linux-386.zip` from gobuilder. So it doesn't actually build an 0.3.x image. We'd have to build the image manually with current master's good dockerfile and push that, which doesn't seem to be allowed for automatic builds.

lgierth replied:
Adding a tiny note about this.

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+The Docker image `ipfs/go-ipfs` is built from master and thus also part of v04x.
+
+ipfs.io provides two essential services to the community,
+which are affected by this: the gateway and the default bootstrappers.
```
"the public http-to-ipfs gateway"

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+
+## How do I update go-ipfs to 0.4.0?
+
+You can use the [ipfs-update tool][ipfs-update] to update go-ipfs.
```
You can: 
- use the ipfs-update tool to update go-ipfs **(recommended)**
- download [the new go-ipfs binaries][go-ipfs-bins] and install them.
- download [the new source](https://github.com/ipfs/go-ipfs) and [build it yourself](https://github.com/ipfs/go-ipfs#build-from-source).

Please note that installing with `go get github.com/ipfs/go-ipfs` does not work at this time. We are experimenting with our [ipfs-based package manager](https://github.com/whyrusleeping/gx).

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+
+## The public gateways at ipfs.io
+
+For the time being, https://ipfs.io uses both v03x and v04x to service requests.
```
add: "Content available in either network will be served just fine."

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+If the first response is not successful (4xx/5xx, connection errors),
+it is discarded and the second response is served, regardless of its status code.
+
+We're using the [multireq proxy][multireq] to accomplish this multiplexing behaviour.
```
Add:
"`multireq` is a new tool, so if you notice any weirdness with gateway requests, [please let us know](https://github.com/ipfs/infrastructure/issues/)."

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+[multireq]: https://github.com/whyrusleeping/multireq
+
+
+## The default bootstrappers
```
maybe: 

## Solarnet: the default bootstrappers

lgierth replied:
Oh yep that's a good occasion to subtly highlight what solarnet actually is :)

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+If you're running an 0.3.x node, it won't be able to communicate
+with any node which has updated to 0.4.x, let alone exchange data.
+You can access data added by 0.4.0 nodes using the public gateway.
+Likewise, data added by your 0.3.x node is still available on the public gateway.
```
If you rely on the public gateways, we ask you to please upgrade to 0.4.0 as soon as you can. You can subscribe [to this issue](https://github.com/ipfs/go-ipfs/issues/2334) to be notified when 0.4.0 is released.

jbenet replied:
If you're running an 0.4.0 node, all is well in the world :sunglasses: You don't need to do anything.

Thank you for bearing with this important change. We seek to provide a smooth transition for everyone, and wish we didn't have to bother you with this at all.

jbenet commented on src/9-v04x-migration/index.md:
```diff
@@ -0,0 +1,92 @@
+which contains lots of great changes and enhancements.
+
+There are breaking changes
+which prevent 0.4.x nodes from communicating with 0.3.x nodes.
```
I think the intro needs some acknowledgement that breaking changes suck, that we take them seriously, and that we've a smooth upgrade path.

> We know breaking changes are painful. We avoid them. In this case, the improvement makes IPFS substantially more upgradable. It's one of those things that "should've been different" from the start. Having the freedom to improve the core protocol is why IPFS is still in alpha phase.
>
> Nevertheless, we take service disruption very seriously even in these early days. We know thousands of developers use IPFS, and hundreds of thousands of people rely on IPFS links to our gateway. So we planned a smooth upgrade path.
>
> The breaking change is at the wire protocol level, which means there will be, and already are, two separate networks. ...

lgierth replied:
> I think the intro needs some acknowledgement that breaking changes suck, that we take them seriously, and that we've a smooth upgrade path.

That's excellent and the basic tone that I couldn't come up with :)

lgierth commented:
Okay addressed feedback

- preview: https://ipfs.io/ipfs/QmYWYq4z5MspeuDCYAyWqko5sG2z4Bx3FqZvMkVgt8RVQq/9-v04x-migration/
- diff: https://github.com/ipfs/blog/commit/4b188000d6a00a992b2575a0d4452bea2ea9d452

jbenet commented:
LGTM.

jbenet commented:
thanks @lgierth!
